### ISSUE TITLE
Training hparams and model evaluation details.

### ISSUE THREAD

**Issue by jwkirchenbauer** [lm-sys/llm-decontaminator#6]
Hi again! 🙂

I now have a series of additional question about your setup, but more focused on the training and performance evaluation.

Is it possible to release the finetuning hyperparameters for the MMLU model trained on the contaminated data that gets very strong performance? Something similar to this [finetune.sh](https://github.com/lm-sys/llm-decontaminator/blob/f1score/finetune.sh) script, which is for gsm8k. Also can you clarify how you evaluated the model performance/accuracy on the test tasks? (not the detection performance in the f1score files) 

I am asking because I noticed that this contaminated training setup seems a bit sensitive to hyperparameters. It would be nice to see a version of the finetune.sh with hyperparameters that achieves 100% accuracy when training on the exact MMLU test questions, and another set of hyperparams (possibly the same) that achieves gpt4 level performance when training on the ngram overlap rejection-sampled paraphrase for each MMLU test question. Basically the training settings for Table2, since section 5.1 doesn't contain quite enough information to reproduce them (even if I had your paraphrased question data accessible, which I don't. would also be nice if you released this!).

Finally, while this wont be visible in the code directly without data preparation code or the raw data, I think it matters how you are presenting the examples to the model at finetuning time and how you are scoring its answers: 
- Do you only compute loss on the one letter target "B." or "D." or do you train on the whole answer like "B. this is the correct answer"? 
- Do you train with a small batch size? do you pack multiple examples into a single sequence? `batch=[xxxyxxyxxxxy, xxyxxxxxyxxy]` or is it one per row like `batch=[xxxy,xxxxy,xxxy,xxxxxy,xxxy,xxy]`
- Did you evaluate the performance of the models on MMLU using the lm-eval-harness method? (loss based ranking) Or did you compute accuracy in some other way?
- How was the performance on other tasks impacted when training on the MMLU test questions and their paraphrases. Did the model's hellaswag, arc, etc. scores remain similar? Or were they heavily impacted by "overfitting" to the MMLU questions over the 16 epochs?

**Comment by andy-yang-1** [comment]
@jwkirchenbauer I will provide a doc for finetuning!

**Comment by andy-yang-1** [comment]
@jwkirchenbauer Thank you very much for your feedback! 

> Is it possible to release the finetuning hyperparameters for the MMLU model trained on the contaminated data that gets very strong performance

Regarding how to fine-tune and some of the hyperparameters, I have provided details [here](https://github.com/lm-sys/llm-decontaminator/tree/main/train#fine-tune). The type of GPU might affect some of the hyperparameters; I have 4 H800 cards, and some settings might differ. The key factor is the epoch, so maintaining the same epochs is fine. On MMLU, I trained 16-32 epochs, while on HumanEval and GSM-8k, I trained 50 epochs. 

> Do you only compute loss on the one letter

The loss calculation is based on the entire sample, not just a single letter. 

> Do you pack multiple examples into a single sequence

I did not pack multiple examples into a single sequence, and I'm not sure how this would impact performance. I will try it out when I have the opportunity.

> Did you evaluate the performance of the models on MMLU using the lm-eval-harness method

The method of evaluation I used is also provided [here](https://github.com/lm-sys/llm-decontaminator/tree/main/train#evaluation). 

> How was the performance on other tasks impacted when training on the MMLU test questions and their paraphrases

We haven't tested if the scores for hellaswag and arc remain consistent, but fine-tuning on rephrased samples can lead to a decline in conversational abilities to some extent. We may try the 'overfitting' problem later.

**Comment by jwkirchenbauer** [comment]
Thanks for the clarifications!

1. The one train script suggests you used a small batch size of 1 with no accumulation. That is correct? (maybe effective batch size of 2 or 4 if you used multiple cards)
2. Okay, so just to be ultra sure, this means the loss was computed on just the answer text or the whole input plus the answer text? i.e. loss computed on:
```
Answer:
B. the answer b
```
versus all the tokens
```
Here is a questions that has the answers?
A. the answer a
B. the answer b
C. the answer c
D. the answer d
Answer:
B. the answer b
```
3. No worries, the sample packing isn't a super standard feature yet for finetuning anyway, it's more common in pretraining. It is unclear what effect it would have on your results, but in some sense it makes the effective batch size a bit larger.
4. Cool! Yes it is interesting to understand whether or not the overall utility of the model is heavily impacted by this very specific scenario of heavy training on a small set of contaminated examples.

Really appreciate the responsiveness! Thanks again for the interesting work you all did

**Comment by andy-yang-1** [comment]
> The one train script suggests you used a small batch size of 1 with no accumulation.

The batch size can be larger, which depends on the size of the GPU memory. On the H800, the `per_device_train_batch_size` is 4.

> loss was computed on just the answer text or the whole input plus the answer text

The loss is based on the entire context.

~~~
Here is a questions that has the answers?
A. the answer a
B. the answer b
C. the answer c
D. the answer d
Answer:
B. the answer b
~~~

> it is interesting to understand whether or not the overall utility of the model is heavily impacted by this very specific scenario of heavy training on a small set of contaminated examples

Yes! I think benchmarks and evaluations are more important than performance, and there is a lot to explore!

**Comment by jwkirchenbauer** [comment]
> The loss is based on the entire context.

Interesting! So that's a bit more like continued language modeling training on the MMLU examples. Any reason you chose to perform training this way? It seems that for instruction tuning/finetuning task data, it is more standard practice to only compute loss on the targets, i.e. for `input=xxxxxxyy` loss is only comparing `yy` to ground truth (though of course the model get all `xxxxxx` as input when it predicts its guess at `yy`).

**Comment by andy-yang-1** [comment]
> Any reason you chose to perform training this way

I think it also works to use `yy` as the loss, and it might even yield better results. However, this would require designing a specific loss for each benchmark, making it impossible to reuse the same training code🤣

**Comment by jwkirchenbauer** [comment]
I think it might yield different results in some cases, but I don't know for sure.
I hear you on the code 🙂
Thanks again!
